tt_metal/impl/device/device.cpp

```diff
@@ -1289,18 +1289,23 @@ bool Device::close() {
     if (not this->initialized_) {
         TT_THROW("Cannot close device {} that has not been initialized!", this->id_);
     }
-    this->deallocate_buffers();
-    watcher_detach(this);
 
     for (const std::unique_ptr<HWCommandQueue> &hw_command_queue : hw_command_queues_) {
         if (hw_command_queue->manager.get_bypass_mode()) {
             hw_command_queue->record_end();
         }
         hw_command_queue->terminate();
     }
+
+    tt_metal::detail::DumpDeviceProfileResults(this, true);
```

Comment:
@pgkeller Profiler data dump is performed after terminate has been sent. 

This is not ideal, there is no guarantee that dispatch BRISCs have reached their FW end when we are reading here.  However, the consequences are not too bad, you might miss FW and KERNEL markers on NCRISC and BRISC (Area encircled in red below). But shorter child calls will be caught. 

<img width="1146" alt="Screenshot 2024-06-04 at 10 02 34 AM" src="https://github.com/tenstorrent/tt-metal/assets/109363418/6ebe7cb4-4dfd-4ceb-9127-0e545db0abc2">

Reply:
Are we ok with a poll on a profiler L1 locations on dispatch cores to flag profiler has finished on those cores?

Reply:
polling L1 from host isn't great, but after terminate is sent shouldn't be a big deal.  maybe usleep a bit first in case a lot of work has been queued so you don't perturb performance you are measuring?  or poll w/ delays between?

Reply:
Yeah this is at the very end of the run and only in profiler builds. I will poll w/ delays, starting with a delay.